gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -1483,7 +1492,43 @@ private Credential getCredential(Configuration config)
       }
     }
 
-    return credential;
+    // If impersonation service account exists, then use current credential to request access token
+    // for the impersonating service account.
+    return getImpersonatedCredential(config, credential).orElse(credential);
+  }
+
+  /**
+   * Generate a {@link Credential} from the internal access token provider based on the service
+   * account to impersonate.
+   */
+  private static Optional<Credential> getImpersonatedCredential(
+      Configuration config, Credential credential) throws IOException {
+    String serviceAccountToImpersonate =
+        IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
+            .withPrefixes(CONFIG_KEY_PREFIXES)
+            .get(config, config::get);
+
+    if (isNullOrEmpty(serviceAccountToImpersonate)) {
+      return Optional.empty();
+    }
+
+    GoogleCloudStorageOptions options =
+        GoogleHadoopFileSystemConfiguration.getGcsFsOptionsBuilder(config)
+            .build()
+            .getCloudStorageOptions();
+    HttpTransport httpTransport =
+        HttpTransportFactory.createHttpTransport(
+            options.getTransportType(),
+            options.getProxyAddress(),
+            options.getProxyUsername(),
+            options.getProxyPassword());
+    GoogleCredential impersonatedCredential =
+        new GoogleCredentialWithIamAccessToken(
+            httpTransport,
+            new CredentialHttpRetryInitializer(credential),
+            serviceAccountToImpersonate,
+            CredentialFactory.GCS_SCOPES);
```

Comment:
I have changes this to `GCS_SCOPES`, PTAL if this is correct.

Reply:
Tested GCS_SCOPES works fine. I think as long as vm has right scope to be able to access IAM APIs, the scope here is more like a declarative instead of imperative notion.